# meta-wgs-ont-pipeline
Pipeline for processing and assembling metagenomic WGS ONT data

## File descriptions
- snakefile: Main pipeline file. Contains directions for assembly.
- meta_pipe.yml: Configuration file with instructions to create the necessary Conda environment.

## Usage
1. Make a local clone of the repo:
```
git clone https://github.com/efeickmann/meta-wgs-ont-pipeline.git \
&& cd meta-wgs-ont-pipeline
```

2. Use the provided config file (```meta_pipe.yml```) to create a Conda environment:
```
conda create -n meta_pipe_env --file meta_pipe.yml
```

3. Open ```snakefile``` in your favorite text editor and edit the following parameters:
- RUN_DATE: The date your ONT data was produced, in the form YYYYMMDD. Used to ID the sequencing run.
- FLOW_CELL: The ID of the flowcell used for the run.
- ONT_MODEL: The chemistry, flowcell type, and basecaller used to generate your data. Should be given in the form ```r1041_min_high_g303``` for data produced with R10.4.1 chemistry on a MINion flowcell and basecalled with Guppy v3.0.3 set to high accuracy basecalling.
- ARCHIVE_DIR: The directory containing the basecalled fastq files from your run.
- BARCODE_FILE: A .tsv file with a list of the samples you'd like to process, along with their barcodes. Of the form
```
sample   barcode
31278   barcode01
...   ...
```
- MIN_NANO_LEN: The minimum read length you'd like to keep.

4. In the terminal, run ```snakemake -c[CORES]```, replacing [CORES] with however many CPUs you have available.

5. Wait.

6. Visualize your assembly quality using the .html summary file generated by QUAST.

## Construction outline:

### 1. Preprocessing with Filtlong
- Relevant parameters: MIN_NANO_LEN
- Recommend MIN_NANO_LEN > 200


### 2. Assembly using Flye

### 3. Polishing (avoiding reliance on Illumina reads)
- Two rounds of Medaka
- Generate polished versions with:
  - Racon

To add:
- Clair3 (actually a variant caller)
- ntEdit
- NextPolish
- FMLRC2

### 4. Obtain consensus between various polishers (final assembly)

### 5. Assessment with QUAST
